text/0000-proc-macro-include.md

```diff
@@ -0,0 +1,175 @@
+
+/// Read the contents of a file as raw bytes and add it to build dependency info.
+///
+/// The build system executing the compiler will know that the file was accessed during compilation,
+/// and will be able to rerun the build when the contents of the file changes.
+///
+/// If the file fails to be read, this is not automatically a fatal error. The proc macro may
+/// gracefully handle the missing file, or emit a compile error noting the missing dependency.
+///
+/// NOTE: some errors may cause panics instead of returning `io::Error`.
+/// We reserve the right to change these errors into `io::Error`s later.
+fn include_bytes<P: AsRef<str>>(path: P) -> Result<Vec<u8>, std::io::Error>;
+```
+
+As an example, consider a potential implementation of [`core::include`](https://doc.rust-lang.org/stable/core/macro.include.html):
+
+```rust
+#[proc_macro]
+pub fn include(input: TokenStream) -> TokenStream {
+    let mut iter = input.into_iter();
+
+    let result = 'main: if let Some(tt) = iter.next() {
+        let TokenTree::Literal(lit) = tt &&
+        let LiteralValue::Str(path) = lit.value() else {
+            Diagnostic::spanned(tt.span(), Level::Error, "argument must be a string literal").emit();
+            break 'main TokenStream::new();
+        }
+
+        match proc_macro::include(&path) {
+            Ok(token_stream) => token_stream,
+            Err(err) => {
+                Diagnostic::spanned(Span::call_site(), Level::Error, format_args!("couldn't read {path}: {err}")).emit();
+                TokenStream::new()
+            }
+        }
+    } else {
+        Diagnostic::spanned(Span::call_site(), Level::Error, "include! takes 1 argument").emit();
+        TokenStream::new()
+    }
+
+    if let Some(_) = iter.next() {
+        Diagnostic::spanned(Span::call_site(), Level::Error, "include! takes 1 argument").emit();
+    }
+
+    result
+}
+```
+
+(RFC note: this example uses unstable and even unimplemented features for clarity.
+However, this RFC in no way requires these features to be useful on its own.)
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+If a file read is unsuccessful, an encoding of the responsible `io::Error` is passed over the RPC bridge.
+If a file is successfully read but fails to lex, `ErrorKind::Other` is returned.
+
+None of these three APIs should ever cause compilation to fail.
+It is the responsibility of the proc macro to fail compilation if a failed file read is fatal.
+
+The author is unsure of the technical details required to implement this in the compiler.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This is more API surface for the `proc_macro` crate, and the `proc_macro` bridge is already complicated.
+Additionally, this is likely to lead to more proc macros which read external files.
+Moving the handling of `include!`-like macros later in the compiler pipeline
+(read: dependent on name resolution)
```

Comment:
Not sure what this means, the standard `include!` already goes through name resolution.

Reply:
I've removed this note. The (unfounded?) fear is that this pushes file inclusion later in the pipeline, which could cause issues. If proc-macro time would be not be significantly later than when `include!` is resolved and evaluated, then I can just strike this particular drawback.